### PR TITLE
Eq 146 and Eq 150 and Supporting Lemma

### PR DESCRIPTION
**Pull request by MohanadAhmed**
# Equations 146 and 150

Side Lemmas:
- `submatrix_succ_above`: Lemma reduces `update_row ` in the row to be deleted by `submatrix` does not matter.
- `cofactor`: the definition of cofactor of element `i j`
- `adjugate_eq_cofactor_transpose `: equivalence between the mathlib definition (using "Cramer Maps") and the cofactor definition.

### CODE REVIEW

File: src/matrix_cookbook/for_mathlib/linear_algebra/matrix/adjugate.lean [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import data.matrix.notation
+
+/-! # Adjugate Matrix : Extra Lemmas -/
+
+open matrix
+open_locale matrix big_operators
+
+lemma submatrix_succ_above  {n : ℕ} {A : matrix (fin n.succ) (fin n.succ) ℂ} {i j} : 
+  (A.update_row i (pi.single j 1)).submatrix i.succ_above j.succ_above = 
+    A.submatrix i.succ_above j.succ_above :=
```
Comment (eric-wieser) [inline src/matrix_cookbook/for_mathlib/linear_algebra/matrix/adjugate.lean:20]: This is true more generally as
```suggestion
lemma submatrix_succ_above {α k l} {n : ℕ} {A : matrix (fin n.succ) k α} (v : k → α) {f : l → k} {i} : 
  (A.update_row i v).submatrix i.succ_above f = A.submatrix i.succ_above f := 
```

**Comment by MohanadAhmed** [comment]
Can we merge this one?

File: src/matrix_cookbook/for_mathlib/linear_algebra/matrix/adjugate.lean [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  (A.update_row i v).submatrix i.succ_above f = A.submatrix i.succ_above f := 
+begin
+  funext r s,
+  rw submatrix_apply,
+  rw submatrix_apply,
+  rw update_row_apply,
+  rw if_eq_of_eq_false,
+  rw eq_iff_iff, split, apply fin.succ_above_ne, 
+  trivial,
+end
```
Comment (eric-wieser) [inline src/matrix_cookbook/for_mathlib/linear_algebra/matrix/adjugate.lean:28]: You missed the key lemma here, `update_row_ne`:
```suggestion
lemma submatrix_update_row_succ_above {α k l} {n : ℕ} {A : matrix (fin n.succ) k α} (v : k → α) {f : l → k} {i} : 
  (A.update_row i v).submatrix i.succ_above f = A.submatrix i.succ_above f := 
ext $ λ r s, (congr_fun (update_row_ne (fin.succ_above_ne i r) : _ = A _) (f s) : _)

lemma submatrix_update_column_succ_above {α k l} {n : ℕ} {A : matrix k (fin n.succ) α} (v : k → α)
  {f : l → k} {i} : 
  (A.update_column i v).submatrix f i.succ_above = A.submatrix f  i.succ_above:= 
ext $ λ r s, update_column_ne (fin.succ_above_ne i s)
```

File: src/matrix_cookbook/for_mathlib/linear_algebra/matrix/adjugate.lean [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+def cofactor {n : ℕ} (A : matrix (fin n.succ) (fin n.succ) ℂ) (i j: fin n.succ)  :=
+  (-1)^(i + j : ℕ) * det (A.submatrix i.succ_above j.succ_above)
+
+/- Lemma should state
+lemma eq_146 (n : ℕ) (A : matrix (fin n.succ) (fin n.succ) ℂ) (i j) :
+  adjugate A j i = cofactor A i j := 
+-/
+lemma adjugate_eq_cofactor_transpose {n : ℕ} (A : matrix (fin n.succ) (fin n.succ) ℂ)
```
Comment (eric-wieser) [inline src/matrix_cookbook/for_mathlib/linear_algebra/matrix/adjugate.lean:37]: Please generalize this to an arbitrary `comm_ring R`

File: src/matrix_cookbook/3_inverses.lean [hunk 4]
```diff
@@ -22,7 +23,29 @@ lemma eq_145 (h : is_unit A.det) : A * A⁻¹ = 1 ∧ A⁻¹ * A = 1 :=
...
 lemma eq_146 (n : ℕ) (A : matrix (fin n.succ) (fin n.succ) ℂ) (i j) :
-  adjugate A j i = (-1)^(i + j : ℕ) * det (A.submatrix i.succ_above j.succ_above) := sorry
+  adjugate A j i = (-1)^(i + j : ℕ) * det (A.submatrix i.succ_above j.succ_above) := 
+begin
+  /- The comment on line 182 of the adjugate file says that the definition of adjugate
+  uses some kind of cramer map to make some things easier. I guess the price we have to pay
+  is to show equivalence of that definiton to the cofactor definition. Eric mentions that
+  mathlib people do not prefer new definitions so we will not use the definition of 
+  cofactor above.
+  The comment can be found here: https://tinyurl.com/4c55v86t
```
Comment (eric-wieser) [inline src/matrix_cookbook/3_inverses.lean:33]: Let's just use the github link, rather than tinyurl.

**Comment by MohanadAhmed** [comment]
@eric-wieser Any thoughts?

File: src/matrix_cookbook/for_mathlib/linear_algebra/matrix/adjugate.lean [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+lemma submatrix_update_row_succ_above {α k l} {n : ℕ} {A : matrix (fin n.succ) k α} 
+  (v : k → α) {f : l → k} {i} : 
+  (A.update_row i v).submatrix i.succ_above f = A.submatrix i.succ_above f := 
+ext $ λ r s, (congr_fun (update_row_ne (fin.succ_above_ne i r) : _ = A _) (f s) : _)
+
+lemma submatrix_update_column_succ_above {α k l} {n : ℕ} {A : matrix k (fin n.succ) α} (v : k → α)
+  {f : l → k} {i} : 
+  (A.update_column i v).submatrix f i.succ_above = A.submatrix f  i.succ_above:= 
+ext $ λ r s, update_column_ne (fin.succ_above_ne i s)
```
Comment (eric-wieser) [inline src/matrix_cookbook/for_mathlib/linear_algebra/matrix/adjugate.lean:28]: These belong in `for_mathlib/data/matrix/notation.lean`

File: src/matrix_cookbook/3_inverses.lean [hunk 6]
```diff
@@ -57,27 +34,9 @@ lemma eq_148 : adjugate A = (adjugate A)ᵀᵀ := rfl
 lemma eq_149 (n : ℕ) (A : matrix (fin n.succ) (fin n.succ) ℂ) :
   det A = ∑ i : fin n.succ, (-1) ^ (i : ℕ) * A i 0 * det (A.submatrix i.succ_above fin.succ) :=
 det_succ_column_zero _
-lemma eq_150 (n : ℕ) (A : matrix (fin n.succ) (fin n.succ) ℂ) : 
-  -- (-1) ^ (i : ℕ) * adjugate A 0 i represents cofactor
```
Comment (eric-wieser) [inline src/matrix_cookbook/3_inverses.lean]: I already established in the previous equations that `adjugate A j i` was the spelling of `Cof(A, i, j)`

File: src/matrix_cookbook/for_mathlib/linear_algebra/matrix/adjugate.lean [hunk 7]
```diff
@@ -17,28 +17,28 @@ import matrix_cookbook.for_mathlib.data.matrix
...
-  simp only [eq_self_iff_true, if_true],
-  conv_lhs {apply_congr, skip, rw pi.single_apply j (1:R) x, 
-  rw mul_ite, rw ite_mul, rw mul_zero, rw zero_mul, },
-  simp only [mul_one, finset.sum_ite_eq', finset.mem_univ, if_true, 
-    neg_one_pow_mul_eq_zero_iff],
-  rw cofactor,
-  rw submatrix_update_row_succ_above,
+  simp_rw [adjugate_apply, det_succ_row _ j, update_row_self, submatrix_update_row_succ_above],
+  rw [fintype.sum_eq_single i (λ h hjk, _), pi.single_eq_same, mul_one],
+  rw [pi.single_eq_of_ne hjk, mul_zero, zero_mul],
```
Comment (eric-wieser) [inline src/matrix_cookbook/for_mathlib/linear_algebra/matrix/adjugate.lean]: The main trick you missed here is that `simp_rw` saves you the need to do `conv` manually.

The other important trick is to use `fintype.sum_eq_single` which deals with all but one of the terms vanishing for you.

File: src/matrix_cookbook/3_inverses.lean (hunk 6 above)
Comment (MohanadAhmed) [reply src/matrix_cookbook/3_inverses.lean]: I assume you mean equation 149 right?
I think it is better to put the definition of cofactor somewhere and use it. If I understand correctly you are "mentally replacing" `cofactor A i j` everywhere with `adjugate A j i` then doing the proofs. This is of course valid but makes the code look "strained" and leads to the statement that is true about any matrix not only `adjugate A`
`lemma eq_148 : adjugate A = (adjugate A)ᵀᵀ := rfl`

Comment (eric-wieser) [reply src/matrix_cookbook/3_inverses.lean]: Whatever definition you pick, you're going to end up with one proof that's true by `rfl`. Pragmatically there's very little point adding `cofactor A i j` to mathlib if we already have `adjugate A j i`. It doesn't matter that they're defined in different ways; if they're mathematically equal then it's best to have one definition and a lemma that rewrites it to the other implementation.

**Comment by eric-wieser** [comment]
Thanks!

**Pull request merged**